File: src/main/java/com/example/javafx/calculator/Calculator.java
package com.example.javafx.calculator;

import java.math.BigDecimal;

import static java.math.BigDecimal.ZERO;

public class Calculator {
    private final Display display;

    private boolean lastButtonWasDigit;
    private String operator = "+";

    private BigDecimal a = ZERO;
    private BigDecimal b = ZERO;

    public Calculator(Display display) {
        this.display = display;
    }

    public BigDecimal getA() {
        return a;
    }

    public BigDecimal getB() {
        return b;
    }

    public boolean isLastButtonWasDigit() {
        return lastButtonWasDigit;
    }

    public String getOperator() {
        return operator;
    }

    public BigDecimal getNumber() {
        return new BigDecimal(display.getDisplayNumber().replace(',', '.'));
    }

    public void setNumber(BigDecimal number) {
        display.setDisplayNumber(String.valueOf(number));
    }

    public void digit(String digit) {
        if (lastButtonWasDigit) {
            display.setDisplayNumber(display.getDisplayNumber() + digit);
        } else {
            display.setDisplayNumber(digit);
        }
        lastButtonWasDigit = true;
    }

    public void clear() {
        setNumber(ZERO);
        lastButtonWasDigit = false;
        operator  = "+";
        a = ZERO;
        b = ZERO;
    }

    public void negate() {
        BigDecimal newNumber = getNumber().negate();
        setNumber(newNumber);
    }


    public void sqrt() {
        double newNumber = Math.sqrt(getNumber().doubleValue());
        setNumber(new BigDecimal(newNumber));
    }

    public void comma() {
        if (!display.getDisplayNumber().contains(",")) {
            display.setDisplayNumber(display.getDisplayNumber() + ",");
        }
        lastButtonWasDigit = true;
    }

    public void fractionOne() {
        //TODO to implement number = 1 / number
    }

    public void percent() {
        //TODO implement me
    }

    public void operator(String operator) {
        if (lastButtonWasDigit){
            b = getNumber();
            calc();
        }
        this.operator = operator;
    }

    public void enter() {
        if (lastButtonWasDigit) {
            b = getNumber();
        }
        calc();
    }

    private void calc() {
        switch (operator) {
            case "+":
                setNumber(a.add(b));
                break;
            case "-":
                setNumber(a.subtract(b));
                break;
            case "*":
                setNumber(a.multiply(b));
                break;
            case "/":
                setNumber(a.divide(b));
                break;
            default:
                throw new IllegalStateException("Unknown operator " + operator);
        }
        a = getNumber();
        lastButtonWasDigit = false;
    }

}

File: src/main/java/com/example/javafx/calculator/Calculator.kt
package com.example.javafx.calculator

import java.math.BigDecimal

import java.math.BigDecimal.ZERO

class Calculator(private val display: Display) {

    var isLastButtonWasDigit: Boolean = false
    var operator = "+"

    var a = ZERO
    var b = ZERO

    var number: BigDecimal
        get() {
            return BigDecimal(display.displayNumber.replace(',', '.'))
        }
        set(number) {
            display.displayNumber = number.toString()
        }

    fun digit(digit: String) {
        if (isLastButtonWasDigit) {
            display.displayNumber = display.displayNumber + digit
        } else {
            display.displayNumber = digit
        }
        isLastButtonWasDigit = true
    }

    fun clear() {
        number = ZERO
        isLastButtonWasDigit = false
        operator = "+"
        a = ZERO
        b = ZERO
    }

    fun negate() {
        val newNumber = number.negate()
        number = newNumber
    }


    fun sqrt() {
        val newNumber = Math.sqrt(number.toDouble())
        number = BigDecimal(newNumber)
    }

    fun comma() {
        if (!display.displayNumber.contains(",")) {
            display.displayNumber = display.displayNumber + ","
        }
        isLastButtonWasDigit = true
    }

    fun fractionOne() {
        //TODO to implement number = 1 / number
    }

    fun percent() {
        //TODO implement me
    }

    fun operator(operator: String) {
        if (isLastButtonWasDigit) {
            b = number
            calc()
        }
        this.operator = operator
    }

    fun enter() {
        if (isLastButtonWasDigit) {
            b = number
        }
        calc()
    }

    private fun calc() {
        when (operator) {
            "+" -> number = a.add(b)
            "-" -> number = a.subtract(b)
            "*" -> number = a.multiply(b)
            "/" -> number = a.divide(b)
            else -> throw IllegalStateException("Unknown operator " + operator)
        }
        a = number
        isLastButtonWasDigit = false
    }

}
